### PR TITLE
feat: Add the ability to manage git server credentials

### PR DESCRIPTION
**Pull request by JPinkney**
### What does this PR do?

This PR adds the ability for the devworkspace operator to automatically add git server configurations into the git config. It also refactors the git provisioning to make it a bit simpler, seperates gitconfigs and git user credentials, and adds some tests. This functionality is enabled with the following label:
```
labels:
    "controller.devfile.io/git-tls-credential": "true"
```
E.g.
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: tls-configmap-test
  namespace: devworkspace-controller
  labels:
    "controller.devfile.io/git-tls-credential": "true"
  annotations:
    controller.devfile.io/mount-path: /home/theia/tls
data:
  host: https://github.com
  certificate: asdfasdfasdfasdfasdfasdfasdfasdf
```

### What issues does this PR fix or reference?
Part of https://github.com/devfile/devworkspace-operator/issues/613

### Is it tested? How?

Tested manually and have made sure the gitconfig is in the correct format for now. I'm still trying to see how I can actually test these changes against a git server.

```
make docker install
cat <<EOF | oc apply -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: tls-configmap-test
  namespace: devworkspace-controller
  labels:
    "controller.devfile.io/git-tls-credential": "true"
    "controller.devfile.io/watch-configmap": "true"
  annotations:
    controller.devfile.io/mount-path: /home/theia/tls
data:
  host: https://github.com
  certificate: asdfasdfasdfasdfasdfasdfasdfasdf
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: tls-configmap-test2
  namespace: devworkspace-controller
  labels:
    "controller.devfile.io/git-tls-credential": "true"
    "controller.devfile.io/watch-configmap": "true"
  annotations:
    controller.devfile.io/mount-path: /home/theia/tls2/
data:
  host: https://github.com
  certificate: abcdefg
---
kind: Secret
apiVersion: v1
metadata:
  name: git-credentials-secret
  namespace: devworkspace-controller
  annotations:
    controller.devfile.io/mount-path: /home/theia/.git-credentials/
  labels:
    controller.devfile.io/git-credential: 'true'
    "controller.devfile.io/watch-secret": "true"
data:
  credentials: asdasdasdasdasdasdasdasdasdasdasdasd
type: Opaque
---
kind: Secret
apiVersion: v1
metadata:
  name: git-credentials-secret2
  namespace: devworkspace-controller
  annotations:
    controller.devfile.io/mount-path: /home/theia/.git-credentials/
  labels:
    controller.devfile.io/git-credential: 'true'
    "controller.devfile.io/watch-secret": "true"
data:
  credentials: asdasdasdasdasdasdasdasdasdasdasdasd
type: Opaque
EOF
oc apply -f samples/theia-next.yaml
```

Then observe that everything is in the correct place. I.e. the certificates are mounted to `/home/theia/tls/certificate` and `/home/theia/tls2/certificate`. The gitconfig is in `/etc/gitconfig` and has the correct format

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

**Review by amisevsk — COMMENTED** [review]
Generally LGTM, I'll test tomorrow

**Review by amisevsk — APPROVED** [review]
:+1:

File: pkg/constants/metadata.go [hunk 1]
```diff
@@ -36,6 +36,14 @@ const (
 	// see https://git-scm.com/docs/git-credential-store#_storage_format for more details
 	DevWorkspaceGitCredentialLabel = "controller.devfile.io/git-credential"
 
+	// DevWorkspaceGitTLSLabel is the label key to specify if the configmap is credentials for accessing a git server.
+	// Configmap must contain the following data:
+	// certificate: the certificate used to access the git server in Base64 ASCII
+	// You can also optionally define the git host.
+	// host: the url of the git server
+	// If the git host is not defined then the certificate will be used for all http repositories.
```
Comment (sleshchenko) [inline pkg/constants/metadata.go:44]:
```suggestion
	// If the git host is not defined then the certificate will be used for all git/ssh repositories.
```
?

File: pkg/constants/metadata.go [hunk 2]
```diff
@@ -36,6 +36,14 @@ const (
 	// see https://git-scm.com/docs/git-credential-store#_storage_format for more details
 	DevWorkspaceGitCredentialLabel = "controller.devfile.io/git-credential"
 
+	// DevWorkspaceGitTLSLabel is the label key to specify if the configmap is credentials for accessing a git server.
+	// Configmap must contain the following data:
+	// certificate: the certificate used to access the git server in Base64 ASCII
```
Comment (sleshchenko) [inline pkg/constants/metadata.go:41]: Just wonder if:
- it makes sense to make it clear that it's private cert via configmap item.
- add an ability optionally to add public cert. Then users are able to grab it from there or dashboard/che server/theia could upload it to git provider. Probably it's already possible but let's make it clear

**Review by sleshchenko — APPROVED** [review]

**Pull request merged**